oak_functions/loader/src/server.rs

```diff
@@ -14,65 +14,207 @@
+            })?;
+        self.response_bytes = response;
+        Ok(())
+    }
 }
 
+impl wasmi::Externals for WasmInterface {
+    /// Invocation of a host function specified by its registered index. Acts as a wrapper for
+    /// the relevant native function, just:
+    /// - checking argument types (which should be correct as `wasmi` will only pass through those
+    ///   types that were specified when the host function was registered with `resolv_func`).
+    /// - mapping resulting return/error values.
+    fn invoke_index(
+        &mut self,
+        index: usize,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Option<wasmi::RuntimeValue>, wasmi::Trap> {
+        match index {
+            READ_REQUEST => map_host_errors(self.read_request(
+                args.nth_checked(0)?,
+                args.nth_checked(1)?,
+                args.nth_checked(2)?,
+            )),
+            WRITE_RESPONSE => {
+                map_host_errors(self.write_response(args.nth_checked(0)?, args.nth_checked(1)?))
+            }
+            _ => panic!("Unimplemented function at {}", index),
+        }
+    }
+}
+
+impl wasmi::ModuleImportResolver for WasmInterface {
+    fn resolve_func(
+        &self,
+        field_name: &str,
+        signature: &wasmi::Signature,
+    ) -> Result<wasmi::FuncRef, wasmi::Error> {
+        oak_resolve_func(field_name, signature)
+    }
+}
 /// Encapsulates the state of a Wasm invocation for a single user request.
 struct WasmState {
     instance: wasmi::ModuleRef,
-    #[allow(dead_code)]
-    memory: wasmi::MemoryRef,
-    #[allow(dead_code)]
-    request_bytes: Vec<u8>,
-    response_bytes: Vec<u8>,
+    abi: WasmInterface,
 }
 
 impl WasmState {
     fn new(module: &wasmi::Module, request_bytes: Vec<u8>) -> anyhow::Result<WasmState> {
-        // TODO(#1919): Make request body available to the Wasm module via ABI functions.
-        // TODO(#1919): Get the actual response from the Wasm module via ABI functions.
-        let instance = wasmi::ModuleInstance::new(module, &wasmi::ImportsBuilder::default())
-            .context("failed to instantiate Wasm module")?
-            .assert_no_start();
-        let memory = instance
+        let mut abi = WasmInterface::new(request_bytes);
+        let instance = wasmi::ModuleInstance::new(
+            module,
+            &wasmi::ImportsBuilder::new().with_resolver("oak_functions", &abi),
+        )
+        .context("failed to instantiate Wasm module")?
+        .assert_no_start();
+
+        abi.memory = instance
             .export_by_name("memory")
             .context("could not find Wasm `memory` export")?
             .as_memory()
-            .cloned()
-            .context("could not interpret Wasm `memory` export as memory")?;
-        Ok(WasmState {
-            instance,
-            memory,
-            request_bytes,
-            response_bytes: "Welcome to Oak Functions\n".as_bytes().to_vec(),
-        })
+            .cloned();
```

Comment:
I think this should still try to unwrap / context the memory, and if it fails it should return early. It does not matter that it returns an Option and you need an Option, really here you should be setting the Option to Some(memory), and if the memory is not there it should stop as soon as possible, not fail later on with a confusing error message.